### PR TITLE
Create Customer Repository

### PR DESCRIPTION
**Pull request by PinkBlink**
Create CustomerRepository according to ScooterRepository
All tests are passed

### CODE REVIEW

File: src/test/java/CustomerRepositoryTest.java [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+public class CustomerRepositoryTest extends BaseTest {
+    private static final Customer CUSTOMER = CustomerTestData.aCustomer().build();
+
+    private  final CustomerRepository customerRepository = new CustomerRepositoryImpl(ConnectionPool.INSTANCE);
+
+    @Test
+    public void saveSetCreateAndUpdateTime() {
+        final var savedCustomer = customerRepository.save(CUSTOMER);
+
+        assertTrue(savedCustomer.getCreateTime().equals(savedCustomer.getUpdateTime())
```
Comment (viktorrmedvedev) [inline src/test/java/CustomerRepositoryTest.java]: assertEquals

Comment (viktorrmedvedev) [reply src/test/java/CustomerRepositoryTest.java]: Assertions are happening on separateLines
actually the correct way to write this test is:

final var customer = aCustomer().build();
final var savedCustomer = customerRepository.save(customer);

final var customerFromDb = customerRepository.get(customer.id);
assertEquals(customerFromDb, savedCustomer);
assertNotNull(customerFromDb.getCreateTime());
assertNotNull(customerFromDb.getUpdateTime());

File: src/test/java/CustomerRepositoryTest.java [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+public class CustomerRepositoryTest extends BaseTest {
+    private static final Customer CUSTOMER = CustomerTestData.aCustomer().build();
+
+    private  final CustomerRepository customerRepository = new CustomerRepositoryImpl(ConnectionPool.INSTANCE);
+
+    @Test
+    public void saveSetCreateAndUpdateTime() {
```
Comment (viktorrmedvedev) [inline src/test/java/CustomerRepositoryTest.java]: savesCustomerToDb

File: src/test/java/CustomerRepositoryTest.java [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+import java.util.NoSuchElementException;
+import java.util.UUID;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+public class CustomerRepositoryTest extends BaseTest {
+    private static final Customer CUSTOMER = CustomerTestData.aCustomer().build();
```
Comment (viktorrmedvedev) [inline src/test/java/CustomerRepositoryTest.java]: I would prefer to not create static constructions for tests as it's pretty much useless

File: src/test/java/CustomerRepositoryTest.java [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+        final var customerFromDatabase = customerRepository.findById(CUSTOMER.getId()).get();
+
+        assertEquals(CUSTOMER, customerFromDatabase);
+    }
+
+    @Test
+    public void findByIdReturnsEmptyOptional() {
+        final var maybeCustomer = customerRepository.findById(UUID.randomUUID());
+
+        assertThrows(NoSuchElementException.class, maybeCustomer::get);
```
Comment (viktorrmedvedev) [inline src/test/java/CustomerRepositoryTest.java]: what a weird way :D

better use assertEquals(maybeCustomer, Optional.empty());

File: src/test/java/CustomerRepositoryTest.java [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+    @Test
+    public void findByEmailReturnsEmptyOptional() {
+        final var maybeCustomer = customerRepository.findByEmail("nothing@else.matters");
+
+        assertThrows(NoSuchElementException.class, maybeCustomer::get);
+    }
+
+    @Test
+    public void findAllReturnsFullyList() {
+        final var customerList = List.of(
```
Comment (viktorrmedvedev) [inline src/test/java/CustomerRepositoryTest.java:70]: why do you need 4 of them :) 
2 is enough

File: src/test/java/CustomerRepositoryTest.java [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+        customerRepository.update(updatedCustomer);
+        final var customerFromDB = customerRepository.findById(updatedCustomer.getId()).get();
+
+        assertEquals(newName, customerFromDB.getName());
+    }
+
+    @Test
+    public void updatesSetUpdateTime() {
+        Customer savedCustomer = customerRepository.save(CUSTOMER);
+        Customer updatedCustomer = customerRepository.update(savedCustomer.toBuilder().updateTime(null).build());
```
Comment (viktorrmedvedev) [inline src/test/java/CustomerRepositoryTest.java]: final var

File: src/test/java/CustomerRepositoryTest.java [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+        final var customerFromDatabase = customerRepository.findByEmail(CUSTOMER.getEmail()).get();
+
+        assertEquals(CUSTOMER, customerFromDatabase);
+    }
+
+    @Test
+    public void findByEmailReturnsEmptyOptional() {
+        final var maybeCustomer = customerRepository.findByEmail("nothing@else.matters");
+
+        assertThrows(NoSuchElementException.class, maybeCustomer::get);
```
Comment (viktorrmedvedev) [inline src/test/java/CustomerRepositoryTest.java]: normally you won't need to write tests for cases when repository returns empty optional

File: src/test/java/CustomerRepositoryTest.java [hunk 8]
```diff
@@ -6,71 +6,57 @@
...
-import java.util.UUID;
 
 import static org.junit.jupiter.api.Assertions.assertEquals;
 import static org.junit.jupiter.api.Assertions.assertNotNull;
 import static org.junit.jupiter.api.Assertions.assertThrows;
 import static org.junit.jupiter.api.Assertions.assertTrue;
 
 public class CustomerRepositoryTest extends BaseTest {
-    private static final Customer CUSTOMER = CustomerTestData.aCustomer().build();
+    private final Customer customer = CustomerTestData.aCustomer().build();
```
Comment (viktorrmedvedev) [inline src/test/java/CustomerRepositoryTest.java]: why won't you build a separate customer for each test?
IMO it's easier to see what's going on inside the test when you set up all test data inside of it

File: src/test/java/CustomerRepositoryTest.java [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+import org.riders.sharing.repository.impl.CustomerRepositoryImpl;
+
+import java.util.List;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+public class CustomerRepositoryTest extends BaseTest {
```
Comment (viktorrmedvedev) [inline src/test/java/CustomerRepositoryTest.java]: You can implement CustomerTestData interface and just call aCustomer().build() in every test to build separate customer for every test.

`public class CustomerRepositoryTest extends BaseTest implements CustomerTestData`

File: src/test/java/ScooterRepositoryTest.java [hunk 10]
```diff
@@ -18,16 +18,16 @@
...
-    private final ScooterRepository scooterRepository = new ScooterRepositoryImpl(connectionPool);
+
+    private final ScooterRepository scooterRepository = new ScooterRepositoryImpl(ConnectionPool.INSTANCE);
 
     @Test
     public void saveSetCreateAndUpdateTime() {
         final var savedScooter = scooterRepository.save(SCOOTER_1);
 
-        assertTrue(savedScooter.getCreateTime().equals(savedScooter.getUpdateTime()) && SCOOTER_1.equals(savedScooter));
+        assertTrue(savedScooter.getCreateTime().equals(savedScooter.getUpdateTime())
```
Comment (viktorrmedvedev) [inline src/test/java/ScooterRepositoryTest.java:29]: adapt ScooterRepositoryTest accordingly to all the comments for CustomerRepositoryTest

**Review by viktorrmedvedev — APPROVED** [review]
